wireapp/wire-ios-mono
# feat: share logs through Wire - WPB-9135

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9135" title="WPB-9135" target="_blank"><img alt="Story" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10815?size=medium" />WPB-9135</a>  Share logs through Wire
  </td></table>
  <br />
 















This PR introduces a new feature: **log sharing through Wire**

Users now have the option to share logs via their mail provider, or via Wire directly, by choosing which conversations they wish to send their log files to.

### Changes

I removed the `SettingsTechnicalReportViewController` and replaced it with a new module. It uses a **MVVM + Router** pattern. 

I used the existing `ShareViewController` to present a list of conversations to choose from to share the logs. I also created a `ShareableDebugReportView` to represent the log file to be shared. It only displays the file but could be extended further if we ever want to include text.

I added two use cases:
- `FetchShareableConversationsUseCase` 
- `ShareFileUseCase`

I use them in the view model to fetch conversations and share the log files. One of the advantages of using use cases here is that it helps with writing unit tests for the view model. The use cases are also straightforward to test.

I extracted some of the log file generation logic out of the `MFMailComposeViewController+Logs` extension into a new struct `LogFilesProvider`. And added a functionality to generate a URL to the log files archive.

I updated `FileMetaDataGenerator.metadataForFileAtURL` from a static func to a non-static func in order to facilitate mocking and dependency injection. Some class extensions were using the static func: since class extensions cannot have stored properties, I introduced a `shared` property in `FileMetaDataGenerator` to allow the extensions to use it as a singleton. I immediately deprecated the property to also ensure we instantiate the class whenever possible.

### Tests

Added unit / snapshot tests for:
- `SettingsDebugReportViewModel`
- `SettingsDebugReportViewController`
- `FetchShareableConversationsUseCase` 
- `ShareFileUseCase`

### Testing

1. Go to `Settings` > `Advanced` > `Debug Report`
2. Tap `Send Report` button and verify that you're able to send logs via email (or share sheet if email isn't configured)
3. Tap `Share Report Via Wire` button, select the conversation(s) you want to share the logs to, tap send. Verify the logs were sent

### Checklist
